File: 数据结构/平衡树(Treap).java
class int2{
	int x;
	int2(int _x){x=_x;}
	public void set(int _x){x=_x;}
}
class TreapNode{
	int x=0,son=0,rep=0,rnd=0;
	int2 l=new int2(0),r=new int2(0);
	TreapNode(int _x){l.set(0);r.set(0);son=rep=1;rnd=new Random().nextInt(2147483647);x=_x;}
}
class Treap{
	int treesize=0;
	int2 root=new int2(0);
	TreapNode[] t=new TreapNode[2000001];
	Treap() {t[0]=new TreapNode(0);t[0].son=t[0].rep=t[0].rnd=0;}
	public void insert(int x){Insert(root,x);}
	public int delete(int x){return Delete(root,x);}
	public int get_succ(int x){return Get_Succ(root,x);}
	public int get_pred(int x){return Get_Pred(root,x);}
	public int ask_kth(int x){return Ask_Kth(root,x);}
	public int ask_rank(int x){return Ask_Rank(root,x);}
	private void update(int k){t[k].son=t[t[k].l.x].son+t[t[k].r.x].son+t[k].rep;}
	private void right_rotate(int2 k){int tt=t[k.x].l.x;t[k.x].l.set(t[tt].r.x);t[tt].r.set(k.x);update(k.x);update(tt);k.set(tt);}
    private void left_rotate(int2 k){int tt=t[k.x].r.x;t[k.x].r.set(t[tt].l.x);t[tt].l.set(k.x);update(k.x);update(tt);k.set(tt);}
    private void Insert(int2 k,int x)
	{
		if (k.x==0)
		{
			k.set(++treesize);
			t[treesize]=new TreapNode(x);
			return;
		}
		t[k.x].son++;
		if (x==t[k.x].x) {t[k.x].rep++;return;}
		if (x<t[k.x].x) {Insert(t[k.x].l,x);if (t[t[k.x].l.x].rnd<t[k.x].rnd)right_rotate(k);}
		else if (x>t[k.x].x) {Insert(t[k.x].r,x);if (t[t[k.x].r.x].rnd<t[k.x].rnd)left_rotate(k);}
	}
	private int Delete(int2 k,int x)
	{
		if (k.x==0)return 0;
		if (x==t[k.x].x)
		{
			if (t[k.x].rep>1){t[k.x].rep--;t[k.x].son--;return 1;}
			if (t[k.x].l.x==0||t[k.x].r.x==0){k.set(t[k.x].l.x+t[k.x].r.x);return 1;}
			if (t[t[k.x].l.x].rnd<t[t[k.x].r.x].rnd){right_rotate(k);return Delete(k,x);}
			left_rotate(k);return Delete(k,x);
		}
		int res;
		if (x<t[k.x].x){res=Delete(t[k.x].l,x);if (res!=0)t[k.x].son--;return res;}
		res=Delete(t[k.x].r,x);if (res!=0)t[k.x].son--;return res;
	}
	private int Get_Succ(int2 k,int x)
	{
		if (k.x==0)return -1;
		int sv;
		if (x<t[k.x].x) {sv=Get_Succ(t[k.x].l,x);return sv==-1?t[k.x].x:sv;}
		return Get_Succ(t[k.x].r,x);
	}
	private int Get_Pred(int2 k,int x)
	{
		if (k.x==0)return -1;
		int sv;
		if (x>t[k.x].x) {sv=Get_Pred(t[k.x].r,x);return sv==-1?t[k.x].x:sv;}
		return Get_Pred(t[k.x].l,x);
	}
	private int Ask_Kth(int2 k,int x)
	{
		if (k.x==0)return -1;
		if (x<=t[t[k.x].l.x].son)return Ask_Kth(t[k.x].l,x);
		if (x>t[t[k.x].l.x].son+t[k.x].rep)return Ask_Kth(t[k.x].r,x-t[t[k.x].l.x].son-t[k.x].rep);
		return t[k.x].x;
	}
	private int Ask_Rank(int2 k,int x)
	{
		if (k.x==0)return -1;
		if (x==t[k.x].x)return t[t[k.x].l.x].son+1;
		if (x<t[k.x].x)return Ask_Rank(t[k.x].l,x);
		return t[t[k.x].l.x].son+t[k.x].rep+Ask_Rank(t[k.x].r,x);
	}
}

File: 数据结构/平衡树(Treap).cpp
/*
 * 不要在函数里或者main里开Treap<int>T,开不下 放全局的
 * T.clear()清空
 * T.insert(x)插入
 * T.del(x)删除，没有成功删除返回false
 * T.pred(x)返回第一个小于x的数，T.pred(x+1)返回第一个小等于x的数
 * T.succ(x)返回第一个大于x的数，T.succ(x-1)返回第一个大等于x的数
 * T.ask_kth(x)返回第x小的数
 * T.ask_rank(x)返回x的排名是第几小
 */
template<class Tdata>
class Treap {
public:
#define Treap_size 200010 /* 记得改这个东西 */
#define NotFound (Tdata)-1
    struct treap_point {
        int l, r, rep, son, rnd;
        Tdata x;
    } t[Treap_size];
    int treesz, root;
    void update(int k) {
        t[k].son = t[t[k].l].son + t[t[k].r].son + t[k].rep;
    }
    void right_rotate(int &k) {
        int tt = t[k].l;
        t[k].l = t[tt].r;
        t[tt].r = k;
        update(k);
        update(tt);
        k = tt;
    }
    void left_rotate(int &k) {
        int tt = t[k].r;
        t[k].r = t[tt].l;
        t[tt].l = k;
        update(k);
        update(tt);
        k = tt;
    }
    void insert_to_tree(int &k, Tdata x) {
        if (!k) {
            k = ++treesz;
            t[k].x = x;
            t[k].rnd = rand();
            t[k].son = t[k].rep = 1;
            t[k].l = t[k].r = 0;
            return;
        }
        t[k].son++;
        if (x == t[k].x) {
            t[k].rep++;
            return;
        }
        if (x < t[k].x) {
            insert_to_tree(t[k].l, x);
            if (t[t[k].l].rnd < t[k].rnd)right_rotate(k);
        }
        if (x > t[k].x) {
            insert_to_tree(t[k].r, x);
            if (t[t[k].r].rnd < t[k].rnd)left_rotate(k);
        }
    }
    bool delete_from_tree(int &k, Tdata x) {
        if (!k)return false;
        if (x == t[k].x) {
            if (t[k].rep > 1) {
                t[k].rep--;
                t[k].son--;
                return true;
            }
            if (!t[k].l || !t[k].r) {
                k = t[k].l + t[k].r;
                return true;
            }
            if (t[t[k].l].rnd < t[t[k].r].rnd) {
                right_rotate(k);
                return delete_from_tree(k, x);
            }
            left_rotate(k);
            return delete_from_tree(k, x);
        }
        bool res;
        if (x < t[k].x) {
            res = delete_from_tree(t[k].l, x);
            if (res)t[k].son--;
            return res;
        }
        res = delete_from_tree(t[k].r, x);
        if (res)t[k].son--;
        return res;
    }
    Tdata get_succ_in_tree(int k, Tdata x) {
        if (!k) return NotFound;
        Tdata sv;
        if (x < t[k].x) {
            sv = get_succ_in_tree(t[k].l, x);
            return sv == NotFound ? t[k].x : sv;
        }
        return get_succ_in_tree(t[k].r, x);
    }
    Tdata get_pred_in_tree(int k, Tdata x) {
        if(!k) return NotFound;
        Tdata sv;
        if (x > t[k].x) {
            sv = get_pred_in_tree(t[k].r, x);
            return sv == NotFound ? t[k].x : sv;
        }
        return get_pred_in_tree(t[k].l, x);
    }
    Tdata ask_kth_in_tree(int k, int x) {
        if (!k) return NotFound;
        if (x <= t[t[k].l].son) return ask_kth_in_tree(t[k].l, x);
        if (x > t[t[k].l].son + t[k].rep) return ask_kth_in_tree(t[k].r, x - t[t[k].l].son - t[k].rep);
        return t[k].x;
    }
    int ask_rank_in_tree(int k, Tdata x) {
        if (!k) return -1;
        if (x == t[k].x) return t[t[k].l].son + 1;
        if (x < t[k].x) return ask_rank_in_tree(t[k].l, x);
        return t[t[k].l].son + t[k].rep + ask_rank_in_tree(t[k].r, x);
    }
    Treap() {
        treesz = root = 0;
    }
    void clear() {
        treesz = root = 0;
    }
    void insert(Tdata x) {
        insert_to_tree(root, x);
    }
    bool del(Tdata x) {
        return delete_from_tree(root, x);
    }
    Tdata pred(Tdata x) {
        return get_pred_in_tree(root, x);
    }
    Tdata succ(Tdata x) {
        return get_succ_in_tree(root, x);
    }
    Tdata ask_kth(int x) {
        return ask_kth_in_tree(root, x);
    }
    int ask_rank(Tdata x) {
        return ask_rank_in_tree(root, x);
    }
#undef NotFound
#undef Treap_size
};
Treap<int>T;